app/models/manageiq/providers/redhat/connection_manager.rb

```diff
@@ -0,0 +1,227 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'ovirtsdk4'
+
+#
+# This class is a responsible for managing a set of `OvirtSDK4::Connection` objects. Each connection is
+# identified by the `id` attribute of the corresponding EMS, or by `nil` if there is no such EMS created yet.
+#
+# The connections will be created the first time that they are needed.
+#
+# Connections will be closed and created again if any of the options used to create them change.
+#
+# All the connections will be closed when the process finishes.
+#
+# Periodically the manager will obtain the identifiers of the EMSs from the database, and will automatically
+# close the connections that correspond to identifiers that no longer exist.
+#
+class ManageIQ::Providers::Redhat::ConnectionManager
+  #
+  # Returns the singleton instance.
+  #
+  def self.instance
+    @instance ||= new
+  end
+
+  #
+  # Creates a connection manager with an empty set of connections. This class is intended to be used as a
+  # singleton, so use the `instance` class method instead.
+  #
+  # @api private
+  #
+  def initialize
+    # This hash stores the connections that have already been created. The keys of the hash will be the
+    # identifiers of the EMSs, and the values will be instances of the `Entry` class.
+    @registry = {}
+
+    # Load from the configuration the settings that control how often we purge connections that correspond to
+    # EMSs that no longer exist:
+    @purge_interval = settings.purge_interval.to_i_with_method
+    @purge_time = Time.now
+
+    # Make sure that all connections will be closed when the process finishes:
+    at_exit do
+      $rhevm_log.info("Closing all connections before exit.")
+      clear
+    end
+  end
+
+  #
+  # Returns the connection that corresponds to the given EMS identifier and options. The connection
+  # will be created if doesn't exist yet.
+  #
+  # If the connection already exists, but the given options are different to the options that were used to
+  # create it, then it will be closed and created again. This is intended to support updates to the
+  # credentials, for example changes of user names or passwords.
+  #
+  # @param id [Object] The id of the EMS.
+  #
+  # @param opts [Hash] The options that will be used to create the connection if it doesn't exist yet. The possible
+  #   values are the same used in the constructor of the `OvirtSDK4::Connection` object.
+  #
+  # @return [OvirtSDK4::Connection] The connection that matches the given EMS id and options.
+  #
+  def get(id, opts)
```

Comment:
Shouldn't we validate whether the id represents existing ems?

Reply:
I think that isn't necessary. In addition the id may be `nil` if the EMS hasn't been created yet (when testing connection details).